internal/controller/management/serviceaccount/serviceaccounts.go

```diff
@@ -0,0 +1,229 @@
+			// controller label was deleted, hence remove Controller
+			// Permissions for the ServiceAccount.
+			return ctrl.Result{}, r.deleteRoleBindingsForAllProjects(ctx, sa)
+		}
+		logger.Error(err, "Failed to get ServiceAccount")
+		return ctrl.Result{}, err
+	}
+
+	if controllerutil.AddFinalizer(sa, kargoapi.FinalizerName) {
+		if err := r.Update(ctx, sa); err != nil {
+			return ctrl.Result{}, err
+		}
+		logger.Debug("Added finalizer to ServiceAccount", "serviceaccount", sa.Name)
+		return ctrl.Result{}, nil
+	}
+
+	// Handle deletion logic
+	if sa.DeletionTimestamp != nil {
+		logger.Debug("Deleting role bindings for ServiceAccount", "serviceaccount", sa.Name)
+		if err := r.deleteRoleBindingsForAllProjects(ctx, sa); err != nil {
+			return ctrl.Result{}, err
+		}
+		controllerutil.RemoveFinalizer(sa, kargoapi.FinalizerName)
+		if err := r.Update(ctx, sa); err != nil {
+			return ctrl.Result{}, err
+		}
+		logger.Debug("Removed finalizer from ServiceAccount", "serviceaccount", sa.Name)
+		return ctrl.Result{}, nil
+	}
+
+	// Handle creation and update cases
+	if !hasControllerLabel(sa) {
+		logger.Debug("Deleting role bindings for ServiceAccount",
+			"serviceaccount", sa.Name,
+		)
+		return ctrl.Result{}, r.deleteRoleBindingsForAllProjects(ctx, sa)
+	}
+	logger.Debug("Creating role bindings for ServiceAccount",
+		"serviceaccount", sa.Name,
+	)
+	return ctrl.Result{}, r.createRoleBindingForAllProjects(ctx, sa)
+}
+
+// createRoleBindingForAllProjects creates RoleBindings for all projects
+func (r *ServiceAccountReconciler) createRoleBindingForAllProjects(ctx context.Context, sa *corev1.ServiceAccount) error {
+	
+	roleBindingName := fmt.Sprintf("%s-readonly-secrets", sa.Name)
+
+	logger := logging.LoggerFromContext(ctx).WithValues(
+		"serviceaccount", sa.Name,
+		"roleBinding", roleBindingName,
+	)
+	logger.Info("Starting to create RoleBindings for all projects")
+
+	// Get all projects (namespaces)
+	projectList := &kargoapi.ProjectList{}
+	if err := r.Client.List(ctx, projectList); err != nil {
+		return fmt.Errorf("error listing projects: %w", err)
+	}
+
+	logger.Debug("Fetched project list", "projectCount", len(projectList.Items))
+
+	// Iterate over each project and create RoleBinding
+	for _, project := range projectList.Items {
+
+		roleBinding := &rbacv1.RoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      roleBindingName,
+				Namespace: project.Name, // Project name is the namespace name
+			},
+			RoleRef: rbacv1.RoleRef{
+				APIGroup: rbacv1.GroupName,
+				Kind:     "ClusterRole",
+				Name:     "kargo-controller-secrets-readonly",
+			},
+			Subjects: []rbacv1.Subject{
+				{
+					Kind:      "ServiceAccount",
+					Name:      sa.Name,
+					Namespace: r.cfg.KargoNamespace,
+				},
+			},
+		}
+
+		// Create or update the RoleBinding		
+		if err := r.Client.Get(ctx, client.ObjectKey{Name: roleBindingName, Namespace: project.Name}, roleBinding); err != nil {
```

Comment:
From here to the end of the method, you have some unnecessary cyclomatic complexity. i.e. Conditions go too many levels deep to easily keep track of. Please refactor this to take advantage of early returns.

1. `err != nil && !kubeerr.IsNotFound(err)` --> return this unanticipated error
1. `err != nil` --> this must have been a 404. Create, but then return when the create is done.
1. At this point, you can treat things unconditionally as an update -- almost -- you _could_ check if the update is necessary, but it's not the end of the world if you don't.

Reply:
Although... if you _don't_ check whether the update is necessary, then you could rewind a bit here, skip the get, update unconditionally, and deal with the 404 if it happens. Checking whether the update is necessary is really the only reason to have led off with the get.